alternator/streams.cc

```diff
@@ -0,0 +1,184 @@
+        std::ostringstream ss;
+        ss << data;
+#if __cplusplus >= 202002L
+        auto s = ss.view();
+#else
+        auto s = ss.str();
+#endif
+        return v.SetString(s.data(), s.size(), a);
+    }
+};
+
+template<typename ValueType>
+struct rapidjson::internal::TypeHelper<ValueType, utils::UUID>
+    : public from_string_helper<ValueType, utils::UUID>
+{};
+
+namespace alternator {
+
+// stream arn _has_ to be 37 or more characters long. ugh...
+class stream_arn : public utils::UUID {
+public:
+    using UUID = utils::UUID;
+    static constexpr char marker = 'S';
+
+    stream_arn() = default;
+    stream_arn(const UUID& uuid)
+        : UUID(uuid)
+    {}
+    stream_arn(std::string_view v)
+        : UUID(v.substr(1))
+    {
+        if (v[0] != marker) {
+            throw std::invalid_argument(std::string(v));
+        }
+    }
+    friend std::ostream& operator<<(std::ostream& os, const stream_arn& arn) {
+        const UUID& uuid = arn;
+        return os << marker << uuid;
+    }
+    friend std::istream& operator>>(std::istream& is, stream_arn& arn) {
+        std::string s;
+        is >> s;
+        arn = stream_arn(s);
+        return is;
+    }
+};
+
+}
+
+template<typename ValueType>
+struct rapidjson::internal::TypeHelper<ValueType, alternator::stream_arn>
+    : public from_string_helper<ValueType, alternator::stream_arn>
+{};
+
+namespace alternator {
+
+future<alternator::executor::request_return_type> alternator::executor::list_streams(client_state& client_state, service_permit permit, rjson::value request) {
+    _stats.api_operations.list_streams++;
+
+    auto streams_start = rjson::get_opt<stream_arn>(request, "ExclusiveStartStreamArn");
+    auto limit = rjson::get_opt<int>(request, "Limit").value_or(std::numeric_limits<int>::max());
+    auto table = get_table(_proxy, request, false);
+
+    auto& db = _proxy.get_db().local();
+    auto& cfs = db.get_column_families();
+    auto i = cfs.begin();
+    auto e = cfs.end();
+
+    // TODO: the unordered_map here is not really well suited for partial
+    // querying - we're sorting on local hash order, and creating a table
+    // between queries may or may not miss info. But that should be rate,
+    // and we can probably expect this to be a single call.
+    if (streams_start) {
+        i = std::find_if(i, e, [&](const std::pair<utils::UUID, lw_shared_ptr<column_family>>& p) {
+            return p.first == *streams_start;
+        });
+        if (i != e) {
+            ++i;
+        }
+    }
+
+    auto ret = rjson::empty_object();
+    auto streams = rjson::empty_array();
+
+    std::optional<stream_arn> last;
+
+    for (;limit > 0 && i != e; ++i) {
+        auto s = i->second->schema();
+        auto& ks_name = s->ks_name();
+        auto& cf_name = s->cf_name();
+
+        if (!is_alternator_keyspace(ks_name)) {
+            continue;
+        }
+        if (table && ks_name != table->ks_name()) {
```

Comment:
We actually don't have an explicit `cdc::get_log_for_table`, though it would be useful here. So I just opted for iterating tables and finding it. It is not like the number of tables are gonna be millions.

Reply:
But you added the get_base_table() function, wouldn't it make more sense to add the get_log_for_table() instead or even just has_log_table()?
I know there aren't going to be millions of tables, but the CS fan in me gets offended to see an O(N) algorithm where it should be O(1) :-)